src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs

```diff
@@ -283,10 +284,46 @@ HashSet<ITypeSymbol> typeStack
 				}
 			}
 
+			// Inspect the type parameter as part of its host symbol
+			if( hostType != default ) {
+				int ordinal = hostType.IndexOfArgument( typeParameter.Name );
+
+				// If the ordinal is negative then the type parameter isn't
+				// declared in the host symbol type arguments and that should
+				// be impossible.
+				if( ordinal < 0 ) {
+					throw new InvalidOperationException( $"Unexpected type declaration in type '{symbol.ContainingType.Name}'" );
+				}
+
+				ITypeSymbol typeArgument = hostType.TypeArguments[ordinal];
+				ImmutabilityScope argumentScope = typeArgument.GetImmutabilityScope();
+
+				// Must be SelfAndChildren otherwise we might accept something
+				// that is a descendant of [ImmutableBaseClass], which is
+				// potentially mutable.
+				if( argumentScope == ImmutabilityScope.SelfAndChildren ) {
+					return MutabilityInspectionResult.NotMutable();
+				}
+
+				// We also need to check to see if the type parameter is marked
+				// immutable, since that could also provide restriction on the
+				// types immutability.
+
+				ITypeParameterSymbol hostTypeParameter = hostType.TypeParameters[ordinal];
+				ImmutabilityScope parameterScope = hostTypeParameter.GetImmutabilityScope();
+
+				// Must be SelfAndChildren otherwise we might accept something
+				// that is a descendant of [ImmutableBaseClass], which is
+				// potentially mutable.
+				if( parameterScope == ImmutabilityScope.SelfAndChildren ) {
+					return MutabilityInspectionResult.NotMutable();
+				}
+			}
+
 			// We have to walk all base types and interfaces to find the
 			// type param with the same name and examine those to see if
 			// the immutability attribute is present.
-			INamedTypeSymbol currentType = typeParameter.ContainingType;
+			INamedTypeSymbol currentType = hostType;

```

Comment:
This was the wrong line.  The containing type is the ultimate ancestor.  Now that we have hostSymbol we can properly only look at the class that contains the type parameter we're inspecting.